File: apps/ios/MagicMobile/RelayTransport.swift
import Foundation
import MagicMobileOnDevice

/// What OnDeviceMultiplayer needs from a table connection: Game Center or the cross-play relay.
/// Both report only authenticated peer IDs (a GKPlayer ID or a relay-assigned connection ID).
@MainActor
protocol TablePacketTransport: AnyObject {
    var onPacket: (@MainActor @Sendable (Data, String) -> Void)? { get set }
    var onDisconnect: (@MainActor @Sendable (String) -> Void)? { get set }
    var onError: (@MainActor @Sendable (String) -> Void)? { get set }
    var onPacketRejected: (@MainActor @Sendable (String) -> Void)? { get set }
    func send(_ data: Data, to authenticatedPeerID: String) throws
    func disconnect()
}

extension GameKitTransport: TablePacketTransport {}

/// One seat at a relay table, as the relay reports it.
struct RelayPeer: Equatable, Sendable {
    let id: String
    let name: String
    let connected: Bool
}

enum RelayConfiguration {
    static let defaultURL = URL(string: "https://magicmobile-relay.calebjfeliciano.workers.dev")!

    /// Debug builds may point at a local `wrangler dev` (MAGICMOBILE_RELAY_URL).
    static var url: URL {
        #if DEBUG
        if let value = ProcessInfo.processInfo.environment["MAGICMOBILE_RELAY_URL"], let url = URL(string: value) { return url }
        #endif
        return defaultURL
    }
}

/// The cross-play relay (services/table-relay), in GameKitTransport's role for tables that
/// mix iPhones and Android phones. The relay stamps each delivered packet with the sender's
/// relay-assigned ID; that is the only sender identity this transport reports.
@MainActor
final class RelayTransport: TablePacketTransport {
    var onPacket: (@MainActor @Sendable (Data, String) -> Void)?
    var onDisconnect: (@MainActor @Sendable (String) -> Void)?
    var onError: (@MainActor @Sendable (String) -> Void)?
    var onPacketRejected: (@MainActor @Sendable (String) -> Void)?
    var onRoster: ((_ peers: [RelayPeer], _ full: Bool) -> Void)?
    /// This phone lost the relay and is trying to reclaim its seat.
    var onConnectionChanged: ((Bool) -> Void)?

    private(set) var localPeerID: String?
    private(set) var code: String?
    private(set) var peers: [RelayPeer] = []
    private(set) var isFull = false
    private(set) var seats = 0

    private let baseURL: URL
    private let session: URLSession
    private var task: URLSessionWebSocketTask?
    private var name = "Player"
    private var token: String?
    private var closed = false
    private var generation = 0
    private var reconnectAttempts = 0
    private var parts: [String: [String?]] = [:]
    /// Frames written while the socket reconnects; sent in order once the relay returns the seat.
    private var outbox: [String] = []
    private var outboxCharacters = 0
    private var welcomed = false
    private var pingTask: Task<Void, Never>?
    /// Frames stay well under the relay's 1,000,000-character limit after JSON escaping.
    private static let partCharacters = 400_000

    init(baseURL: URL = RelayConfiguration.url) {
        self.baseURL = baseURL
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 15
        session = URLSession(configuration: configuration)
    }

    /// Creates a table and returns its code and the host key only this phone holds.
    func createTable(seats: Int) async throws -> (code: String, hostKey: String) {
        var request = URLRequest(url: baseURL.appendingPathComponent("v1/tables"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["seats": seats])
        let (data, response): (Data, URLResponse)
        do { (data, response) = try await session.data(for: request) }
        catch { throw EngineError.invalidMessage("Could not reach the table service. Check your connection and try again.") }
        let value = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode),
              let code = value?["code"] as? String, let key = value?["hostKey"] as? String else {
            throw EngineError.invalidMessage((value?["message"] as? String) ?? "The table service could not open a table. Try again.")
        }
        return (code, key)
    }

    /// Opens the table's socket: the creator passes its host key, guests only the code.
    func connect(code: String, name: String, hostKey: String? = nil) {
        self.code = code.uppercased(); self.name = name
        open(hostKey: hostKey, resume: nil)
    }

    private func open(hostKey: String?, resume: String?) {
        guard let code else { return }
        generation += 1
        let current = generation
        var components = URLComponents(url: baseURL.appendingPathComponent("v1/tables/\(code)/socket"), resolvingAgainstBaseURL: false)!
        components.scheme = components.scheme == "https" ? "wss" : "ws"
        components.queryItems = [URLQueryItem(name: "name", value: name)]
            + (hostKey.map { [URLQueryItem(name: "key", value: $0)] } ?? [])
            + (resume.map { [URLQueryItem(name: "resume", value: $0)] } ?? [])
        let task = session.webSocketTask(with: components.url!)
        task.maximumMessageSize = 4 * 1024 * 1024
        self.task = task
        task.resume()
        receiveNext(task, generation: current)
        pingTask?.cancel()
        pingTask = Task { @MainActor [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(20))
                guard let self, current == self.generation, !self.closed else { return }
                self.task?.send(.string("{\"t\":\"ping\"}")) { _ in }
            }
        }
    }

    private func receiveNext(_ task: URLSessionWebSocketTask, generation current: Int) {
        task.receive { [weak self] result in
            Task { @MainActor [weak self] in
                guard let self, current == self.generation else { return }
                switch result {
                case .success(.string(let text)):
                    if !self.closed { self.receive(text) }
                    self.receiveNext(task, generation: current)
                case .success(.data(let data)):
                    if !self.closed, let text = String(data: data, encoding: .utf8) { self.receive(text) }
                    self.receiveNext(task, generation: current)
                case .success:
                    self.receiveNext(task, generation: current)
                case .failure:
                    self.dropped(closeCode: task.closeCode.rawValue)
                }
            }
        }
    }

    private func receive(_ text: String) {
        guard let data = text.data(using: .utf8),
              let frame = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = frame["t"] as? String else { return }
        switch type {
        case "welcome":
            welcomed = true
            localPeerID = frame["you"] as? String
            token = frame["token"] as? String
            seats = (frame["seats"] as? Int) ?? seats
            reconnectAttempts = 0
            let queued = outbox
            outbox.removeAll(); outboxCharacters = 0
            for text in queued { task?.send(.string(text)) { _ in } }
            updateRoster(frame["peers"], full: frame["full"] as? Bool == true)
            onConnectionChanged?(true)
        case "roster":
            updateRoster(frame["peers"], full: frame["full"] as? Bool == true)
        case "msg":
            guard let from = frame["from"] as? String, let payload = frame["d"] as? String else { return }
            guard let part = frame["p"] as? [String: Any] else {
                onPacket?(Data(payload.utf8), from); return
            }
            guard let id = part["id"] as? String, let index = part["i"] as? Int, let count = part["n"] as? Int,
                  (1...64).contains(count), (0..<count).contains(index) else { return }
            let key = "\(from)/\(id)"
            var pieces = parts[key] ?? Array(repeating: nil, count: count)
            guard pieces.count == count else { parts[key] = nil; return }
            pieces[index] = payload
            if pieces.allSatisfy({ $0 != nil }) {
                parts[key] = nil
                onPacket?(Data(pieces.compactMap { $0 }.joined().utf8), from)
            } else {
                parts[key] = pieces
            }
        case "gone":
            if let id = frame["id"] as? String { onDisconnect?(id) }
        case "error":
            closed = true
            onError?((frame["message"] as? String) ?? "The table service closed the connection.")
        default:
            break
        }
    }

    private func updateRoster(_ value: Any?, full: Bool) {
        if let rows = value as? [[String: Any]] {
            peers = rows.compactMap { row in
                guard let id = row["id"] as? String else { return nil }
                return RelayPeer(id: id, name: (row["name"] as? String) ?? "Player", connected: row["connected"] as? Bool == true)
            }
        }
        isFull = full
        onRoster?(peers, full)
    }

    /// A dropped socket reclaims the seat with its resume token, with backoff, for about a minute.
    private func dropped(closeCode: Int) {
        task = nil; welcomed = false
        guard !closed else { return }
        guard let resume = token, closeCode != 4400 else {
            closed = true; onError?("Could not connect to the table."); return
        }
        onConnectionChanged?(false)
        guard reconnectAttempts < 12 else { closed = true; onError?("Lost the connection to the table."); return }
        let delay = min(8.0, 0.5 * pow(2.0, Double(min(reconnectAttempts, 4))))
        reconnectAttempts += 1
        let current = generation
        Task { @MainActor [weak self] in
            try? await Task.sleep(for: .seconds(delay))
            guard let self, !self.closed, self.task == nil, self.generation == current else { return }
            self.open(hostKey: nil, resume: resume)
        }
    }

    func send(_ data: Data, to authenticatedPeerID: String) throws {
        guard !closed else { throw EngineError.invalidMessage("The table connection is closed.") }
        guard let text = String(data: data, encoding: .utf8) else { throw EngineError.invalidMessage("Invalid table packet.") }
        let characters = Array(text)
        if characters.count <= Self.partCharacters {
            try write(try frame(to: authenticatedPeerID, text: text, part: nil))
            return
        }
        let count = (characters.count + Self.partCharacters - 1) / Self.partCharacters
        guard count <= 64 else { throw EngineError.messageTooLarge }
        let id = UUID().uuidString
        for index in 0..<count {
            let slice = String(characters[(index * Self.partCharacters)..<min(characters.count, (index + 1) * Self.partCharacters)])
            try write(try frame(to: authenticatedPeerID, text: slice, part: ["id": id, "i": index, "n": count]))
        }
    }

    /// Sends now, or holds the frame until the relay returns this phone's seat.
    private func write(_ frame: String) throws {
        if let task, welcomed { task.send(.string(frame)) { _ in }; return }
        guard outbox.count < 512, outboxCharacters + frame.count <= 8_000_000 else {
            throw EngineError.invalidMessage("Lost the connection to the table.")
        }
        outbox.append(frame); outboxCharacters += frame.count
    }

    private func frame(to peer: String, text: String, part: [String: Any]?) throws -> String {
        var value: [String: Any] = ["t": "send", "to": peer, "d": text]
        if let part { value["p"] = part }
        let data = try JSONSerialization.data(withJSONObject: value)
        return String(decoding: data, as: UTF8.self)
    }

    /// Leaves the table for good; the other phones are told at once.
    func disconnect() {
        guard !closed || task != nil else { return }
        closed = true
        generation += 1
        pingTask?.cancel(); pingTask = nil
        if let task {
            task.send(.string("{\"t\":\"bye\"}")) { _ in }
            task.cancel(with: .normalClosure, reason: nil)
        }
        task = nil; welcomed = false; outbox.removeAll(); outboxCharacters = 0
        onPacket = nil; onRoster = nil; onDisconnect = nil; onError = nil; onPacketRejected = nil; onConnectionChanged = nil
    }
}

File: apps/android/app/src/main/java/io/magicmobile/android/ondevice/RelayTransport.kt
package io.magicmobile.android.ondevice

import android.os.Handler
import android.os.Looper
import io.magicmobile.android.BuildConfig
import io.magicmobile.android.game.EngineError
import io.magicmobile.android.game.J
import io.magicmobile.android.game.array
import io.magicmobile.android.game.bool
import io.magicmobile.android.game.get
import io.magicmobile.android.game.integer
import io.magicmobile.android.game.obj
import io.magicmobile.android.game.string
import io.magicmobile.android.ui.LaunchEnvironment
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.net.URLEncoder
import java.util.UUID
import java.util.concurrent.TimeUnit

/** One seat at a relay table as the relay reports it. */
data class RelayPeer(val id: String, val name: String, val connected: Boolean)

/**
 * The cross-play relay (services/table-relay): the same role GameKitTransport plays for Game
 * Center. The relay assigns this phone's peer ID and stamps every delivered packet with its
 * sender's ID; that is the only sender identity this transport ever reports.
 */
class RelayTransport(private val baseURL: String = defaultURL) {
    var onPacket: ((ByteArray, String) -> Unit)? = null
    var onRoster: ((List<RelayPeer>, Boolean) -> Unit)? = null
    /** A player left for good (or never came back): the match cannot continue. */
    var onDisconnect: ((String) -> Unit)? = null
    var onError: ((String) -> Unit)? = null
    /** This phone lost the relay and is trying to reclaim its seat. */
    var onConnectionChanged: ((Boolean) -> Unit)? = null

    var localPeerID: String? = null; private set
    var code: String? = null; private set
    var peers: List<RelayPeer> = emptyList(); private set
    var isFull = false; private set
    /** Human seats at this table, from the relay's welcome. */
    var seats = 0; private set

    private val main = Handler(Looper.getMainLooper())
    private var socket: WebSocket? = null
    private var name = "Player"
    private var token: String? = null
    private var closed = false
    private var generation = 0
    private var reconnectAttempts = 0
    private val parts = HashMap<String, Array<String?>>()
    /** Frames written while the socket reconnects; sent in order once the relay returns the seat. */
    private val outbox = ArrayList<String>()
    private var outboxChars = 0
    private var welcomed = false

    /** Creates a table on the relay and returns its code and the host key only this phone holds. */
    suspend fun createTable(seats: Int): Pair<String, String> = withContext(Dispatchers.IO) {
        val body = """{"seats":$seats}""".toRequestBody("application/json".toMediaType())
        val response = try { client.newCall(Request.Builder().url("$baseURL/v1/tables").post(body).build()).execute() }
            catch (error: java.io.IOException) { throw EngineError.InvalidMessage("Could not reach the table service. Check your connection and try again.") }
        response.use {
            val value = runCatching { Json.parseToJsonElement(it.body?.string().orEmpty()) }.getOrNull()
            val code = value["code"].string; val key = value["hostKey"].string
            if (!it.isSuccessful || code == null || key == null) {
                throw EngineError.InvalidMessage(value["message"].string ?: "The table service could not open a table. Try again.")
            }
            code to key
        }
    }

    /** Opens the table's socket: the creator passes its host key, guests only the code. */
    fun connect(code: String, name: String, hostKey: String? = null) {
        this.code = code.uppercase(); this.name = name
        open(hostKey = hostKey, resume = null)
    }

    private fun open(hostKey: String?, resume: String?) {
        val code = code ?: return
        val current = ++generation
        val query = buildList {
            add("name=" + URLEncoder.encode(name, "UTF-8"))
            hostKey?.let { add("key=" + URLEncoder.encode(it, "UTF-8")) }
            resume?.let { add("resume=" + URLEncoder.encode(it, "UTF-8")) }
        }.joinToString("&")
        val url = baseURL.replaceFirst("http", "ws") + "/v1/tables/$code/socket?$query"
        socket = client.newWebSocket(Request.Builder().url(url).build(), object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                main.post { if (current == generation && !closed) receive(text) }
            }
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                main.post { if (current == generation) dropped(code) }
            }
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                main.post { if (current == generation) dropped(null) }
            }
        })
    }

    private fun receive(text: String) {
        val frame = runCatching { Json.parseToJsonElement(text) }.getOrNull()?.obj ?: return
        when (frame["t"].string) {
            "welcome" -> {
                welcomed = true
                localPeerID = frame["you"].string; token = frame["token"].string
                seats = frame["seats"].integer?.toInt() ?: seats
                reconnectAttempts = 0
                val queued = outbox.toList()
                outbox.clear(); outboxChars = 0
                queued.forEach { socket?.send(it) }
                updateRoster(frame["peers"], frame["full"].bool == true)
                onConnectionChanged?.invoke(true)
            }
            "roster" -> updateRoster(frame["peers"], frame["full"].bool == true)
            "msg" -> {
                val from = frame["from"].string ?: return
                val data = frame["d"].string ?: return
                val part = frame["p"].obj
                if (part == null) { onPacket?.invoke(data.toByteArray(Charsets.UTF_8), from); return }
                val id = part["id"].string ?: return
                val index = part["i"].integer?.toInt() ?: return
                val count = part["n"].integer?.toInt() ?: return
                if (count !in 1..64 || index !in 0 until count) return
                val key = "$from/$id"
                val pieces = parts.getOrPut(key) { arrayOfNulls(count) }
                if (pieces.size != count) { parts.remove(key); return }
                pieces[index] = data
                if (pieces.all { it != null }) {
                    parts.remove(key)
                    onPacket?.invoke(pieces.joinToString("") { it!! }.toByteArray(Charsets.UTF_8), from)
                }
            }
            "gone" -> frame["id"].string?.let { onDisconnect?.invoke(it) }
            "error" -> {
                closed = true
                onError?.invoke(frame["message"].string ?: "The table service closed the connection.")
            }
        }
    }

    private fun updateRoster(value: J?, full: Boolean) {
        peers = value.array?.mapNotNull { row ->
            val id = row["id"].string ?: return@mapNotNull null
            RelayPeer(id, row["name"].string ?: "Player", row["connected"].bool == true)
        } ?: peers
        isFull = full
        onRoster?.invoke(peers, full)
    }

    /** A dropped socket reclaims the seat with its resume token, with backoff, for about a minute. */
    private fun dropped(code: Int?) {
        socket = null; welcomed = false
        if (closed) return
        val resume = token
        if (resume == null || code == 4400) { closed = true; onError?.invoke("Could not connect to the table."); return }
        onConnectionChanged?.invoke(false)
        if (reconnectAttempts >= 12) { closed = true; onError?.invoke("Lost the connection to the table."); return }
        val delay = minOf(8_000L, 500L shl minOf(reconnectAttempts, 4))
        reconnectAttempts += 1
        main.postDelayed({ if (!closed && socket == null) open(hostKey = null, resume = resume) }, delay)
    }

    fun send(data: ByteArray, peer: String) {
        if (closed) throw EngineError.InvalidMessage("The table connection is closed.")
        val text = String(data, Charsets.UTF_8)
        if (text.length <= partChars) {
            write(frame(peer, text, null))
            return
        }
        // Never split a surrogate pair: iOS rejects a string holding half of one.
        val pieces = ArrayList<String>()
        var start = 0
        while (start < text.length) {
            var end = minOf(text.length, start + partChars)
            if (end < text.length && text[end - 1].isHighSurrogate()) end -= 1
            pieces += text.substring(start, end)
            start = end
        }
        if (pieces.size > 64) throw EngineError.MessageTooLarge
        val id = UUID.randomUUID().toString()
        pieces.forEachIndexed { index, piece ->
            write(frame(peer, piece, JsonObject(mapOf("id" to JsonPrimitive(id), "i" to JsonPrimitive(index), "n" to JsonPrimitive(pieces.size)))))
        }
    }

    /** Sends now, or holds the frame until the relay returns this phone's seat. */
    private fun write(frame: String) {
        val socket = socket
        if (socket != null && welcomed) { socket.send(frame); return }
        if (outbox.size >= 512 || outboxChars + frame.length > 8_000_000) {
            throw EngineError.InvalidMessage("Lost the connection to the table.")
        }
        outbox += frame; outboxChars += frame.length
    }

    private fun frame(peer: String, text: String, part: JsonObject?): String =
        JsonObject(buildMap {
            put("t", JsonPrimitive("send")); put("to", JsonPrimitive(peer)); put("d", JsonPrimitive(text))
            part?.let { put("p", it) }
        }).toString()

    /** Leaves the table for good; the other phones are told at once. */
    fun disconnect() {
        if (closed && socket == null) return
        closed = true
        generation += 1
        socket?.let { it.send("""{"t":"bye"}"""); it.close(1000, "bye") }
        socket = null; welcomed = false; outbox.clear(); outboxChars = 0
        onPacket = null; onRoster = null; onDisconnect = null; onError = null; onConnectionChanged = null
    }

    companion object {
        /** Frames stay well under the relay's 1,000,000-character limit after JSON escaping. */
        private const val partChars = 400_000
        private val client = OkHttpClient.Builder().pingInterval(20, TimeUnit.SECONDS).connectTimeout(15, TimeUnit.SECONDS).build()

        /** Debug builds may point at a local `wrangler dev` (MAGICMOBILE_RELAY_URL). */
        val defaultURL: String get() = LaunchEnvironment["MAGICMOBILE_RELAY_URL"]?.takeIf { BuildConfig.DEBUG }?.trimEnd('/')
            ?: BuildConfig.RELAY_URL
    }
}
